### ISSUE TITLE
WebView stops using disk cache on erase (was: Report of unexpectedly high Focus network data usage)

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/focus-android#1492]
We received a user review where the user felt Focus was using an unreasonable amount of data. Let's investigate!

2 thoughts:
1. We erase the cache every time the erase button is pressed – I wonder how this would affect browser data usage. Should we make this configurable? Most users probably aren't concerned about the cache sticking around (and it probably sticks around for private browsing in other browsers).
2. WebKit won't recreate the folders we delete (note: may not affect temp files) until the process is restarted (cause of #1306 and further investigation required in #1472) so I wonder if this includes the cache. If WebKit can't save cache files, it may cache fewer files since it can only do so in memory and we may use more data to re-retrieve these files.

**Comment by mcomella** [comment]
We should figure out if this is expected behavior (likely 1) or if there's a bug (likely 2). Triage brought up the fact that private browsing doesn't save cached data so we don't need to set the expectation that we would either.

**Comment by mcomella** [comment]
> 1. We erase the cache every time the erase button is pressed – I wonder how this would affect browser data usage. Should we make this configurable? Most users probably aren't concerned about the cache sticking around (and it probably sticks around for private browsing in other browsers).

I think deleting the cache is desired behavior (to erase traces of browsing history) so I think we should do it. If we receive additional reports of data usage, we could reconsider this decision. I verified `WebView.clearCache` works as expected.

> 2. WebKit won't recreate the folders we delete (note: may not affect temp files) until the process is restarted (cause of #1306 and further investigation required in #1472) so I wonder if this includes the cache.

I confirmed that WebView will never use the disk cache after we delete the WebView files until process restart. To do this, I attached the Chrome dev tools, went to the Network pane, and noticed that the "Size" column no longer indicates "disk cache" after we erase.

We shouldn't delete the cache.

**Comment by Sdaswani** [comment]
Thanks for your investigation @mcomella . I'm a little confused by your last comment though.

You first say: 
> I think deleting the cache is desired behavior (to erase traces of browsing history) so I think we should do it. 

But at the bottom of the comment you say:
> We shouldn't delete the cache.

Let me know what I'm misunderstanding. Perhaps the 'it' you refer to in the first quote is not a reference to 'deleting the cache'?

**Comment by mcomella** [comment]
> Let me know what I'm misunderstanding. Perhaps the 'it' you refer to in the first quote is not a reference to 'deleting the cache'?

The first comment refers to clearing the cache via the WebView APIs (which works correctly), whereas the second comments refers to deleting the cache/ directory on disk (which breaks WebView). So it's desirable to clear the cache via the WebView APIs but not to delete the cache/ files on disk (which my PR addresses).

**Comment by Sdaswani** [comment]
OK understood now. By not deleting the cache/files on disk, we are going to save users on bandwidth usage, but do we lose anything from an ad / tracker blocking point of view?

**Comment by pocmo** [comment]
1. Be cautious - WebView is a moving target and Focus is running on top of various versions of WebView/Chromium/Chrome. Behavior you verify on one device could in theory be different on another one.

2. I think I saw some data not getting erased that is not the average browsing cache data, e.g. data stored by service workers. WebViewDataTest might have some hints (ignore the parts about checking for the existence of folders).

**Comment by mcomella** [comment]
> 1. Be cautious - WebView is a moving target and Focus is running on top of various versions of WebView/Chromium/Chrome. Behavior you verify on one device could in theory be different on another one.

Good point. I did not seriously consider this while doing my analysis.

Fallout from this is explored in the PR: https://github.com/mozilla-mobile/focus-android/pull/1570#issuecomment-340611638.

2. I think I saw some data not getting erased that is not the average browsing cache data, e.g. data stored by service workers. WebViewDataTest might have some hints (ignore the parts about checking for the existence of folders).

As per the comment above, I'll revise my PR to delete as much as possible and try to verify this will cover service workers.

> By not deleting the cache/files on disk, we are going to save users on bandwidth usage, but do we lose anything from an ad / tracker blocking point of view?

To summarize, with my (updated ^) changes, after a session ends we're letting WebView delete its cache and its files on disk, rather than us deleting the files on disk ourselves. When we deleted the files on disk ourselves, it created a bug where WebView would no longer use the disk cache, forcing users to use additional bandwidth. The intent of my changes is to fix this bug – i.e. we still clear the cache between sessions.

From an ad / tracker point of view, we're moving to the intended behavior: within a sesison, users' caches will add bits for fingerprinting but this data will be cleared between sessions.

**Comment by mcomella** [comment]
> I think I saw some data not getting erased that is not the average browsing cache data, e.g. data stored by service workers.

Service workers seem to store data in `<data-dir>/app_webview/Service Workers`, which we delete by hand: this is in opposition to the `<data-dir>/cache/` directory, whose behavior I've been modifying.

**Comment by Sdaswani** [comment]
@mcomella left field question here - is there any chance where data is stored can be configured by the OEM, such that different devices will behave differently?

**Comment by mcomella** [comment]
> is there any chance where data is stored can be configured by the OEM, such that different devices will behave differently?

In theory, OEMs can change _anything_. In practice, it'd be very bad for the OEMs to break certain Android APIs we're using such as:
- `Context.getDataDir`
- `Context.getCacheDir`

So I think it's safe to say these APIs/paths will be consistent. WebView should (hopefully) also use these APIs so the path that it stores its root directory should remain consistent.

WebView is updated independent of the system (from Play) and should be consistent for a given version. However, it **can change where it saves specific data across versions** (I am concerned about this) but we haven't found any evidence of this yet.

On delete, we remove WebKit files that aren't white-listed so **we shouldn't leak browsing data (unless they add it to white-listed files/directories)** but some things we fixed by white-listing files may get broken if the paths change and we delete those files anyway.

Is there a specific path you're concerned about, @Sdaswani ? I think it's good to think about this, especially for #1472 – thanks for bringing this up.

**Comment by Sdaswani** [comment]
Nothing specific @mcomella - just wanted to make sure we weren't targeting a solution with assumptions that may be changed be OEMs. :) I agree it's not recommended for them to do this so hopefully we are OK.

**Comment by mcomella** [comment]
Actually, it'd be ideal to QA this:
- Start focus, go to apple.com
- Connect Chrome developer tools (on desktop) to Focus
- Reload apple.com and verify the disk cache is being used
- Click the trash can
- Go to apple.com
- Reconnect Chrome developer tools (it disconnects when we close the WebView)
- Reload apple.com and verify the disk cache is being used

**Comment by npark-mozilla** [comment]
@mcomella actually, when I connect devtools, it says 'No browser is detected'.  It does detect fine when I launch chrome instead of Focus.  Is there a way to make devtool recognize Focus?

**Comment by npark-mozilla** [comment]
Chrome devtool network tab shows that the data is indeed coming from disk cache when "refreshing," even after trash can icon is clicked. Looks good to me.
